processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -232,167 +239,22 @@ private static void checkIndex(int index, int size)
-  }
-
-  private static <T> GenericIndexed<T> createVersionOneGenericIndexed(ByteBuffer byteBuffer, ObjectStrategy<T> strategy)
-  {
-    boolean allowReverseLookup = byteBuffer.get() == REVERSE_LOOKUP_ALLOWED;
-    int size = byteBuffer.getInt();
-    ByteBuffer bufferToUse = byteBuffer.asReadOnlyBuffer();
-    bufferToUse.limit(bufferToUse.position() + size);
-    byteBuffer.position(bufferToUse.limit());
-
-    return new GenericIndexed<T>(
-        bufferToUse,
-        strategy,
-        allowReverseLookup
-    );
-  }
-
-  private static <T> GenericIndexed<T> createVersionTwoGenericIndexed(
-      ByteBuffer byteBuffer,
-      ObjectStrategy<T> strategy,
-      SmooshedFileMapper fileMapper
-  )
-  {
-    if (fileMapper == null) {
-      throw new IAE("SmooshedFileMapper can not be null for version 2.");
-    }
-    boolean allowReverseLookup = byteBuffer.get() == REVERSE_LOOKUP_ALLOWED;
-    int logBaseTwoOfElementsPerValueFile = byteBuffer.getInt();
-    int numElements = byteBuffer.getInt();
-    String columnName;
-
-    List<ByteBuffer> valueBuffersToUse;
-    ByteBuffer headerBuffer;
-    try {
-      columnName = SERIALIZER_UTILS.readString(byteBuffer);
-      valueBuffersToUse = Lists.newArrayList();
-      int elementsPerValueFile = 1 << logBaseTwoOfElementsPerValueFile;
-      int numberOfFilesRequired = getNumberOfFilesRequired(elementsPerValueFile, numElements);
-      for (int i = 0; i < numberOfFilesRequired; i++) {
-        valueBuffersToUse.add(
-            fileMapper.mapFile(GenericIndexedWriter.generateValueFileName(columnName, i))
-                      .asReadOnlyBuffer()
-        );
-      }
-      headerBuffer = fileMapper.mapFile(GenericIndexedWriter.generateHeaderFileName(columnName));
-    }
-    catch (IOException e) {
-      throw new RuntimeException("File mapping failed.", e);
-    }
-
-    return new GenericIndexed<T>(
-        valueBuffersToUse,
-        headerBuffer,
-        strategy,
-        allowReverseLookup,
-        logBaseTwoOfElementsPerValueFile,
-        numElements
-    );
-  }
-
-  public static <T> GenericIndexed<T> read(ByteBuffer buffer, ObjectStrategy<T> strategy, SmooshedFileMapper fileMapper)
-  {
-    byte versionFromBuffer = buffer.get();
-
-    if (VERSION_ONE == versionFromBuffer) {
-      return createVersionOneGenericIndexed(buffer, strategy);
-    } else if (VERSION_TWO == versionFromBuffer) {
-      return createVersionTwoGenericIndexed(buffer, strategy, fileMapper);
-    }
-
-    throw new IAE("Unknown version [%s]", versionFromBuffer);
-  }
-
   @Override
   public Class<? extends T> getClazz()
   {
-    return bufferIndexed.getClazz();
+    return strategy.getClazz();
   }
 
   @Override
   public int size()
   {
-    return bufferIndexed.size();
+    return size;
   }
 
   @Override
   public T get(int index)
   {
-    return bufferIndexed.get(index);
+    return versionOne ? getVersionOne(index) : getVersionTwo(index);
```

Comment:
I post benchmark results above: https://github.com/druid-io/druid/pull/3951#discussion_r101880981.
It compares only version 1 (before and after), but version 2 won't be very different just because of this ternary in `get()`, because there is no difference for JIT which branch to take, it will more up (invert condition for this) & inline the hotter branch anyway.